### PR TITLE
fix Issue 19384 - [Codegen] Address of stack temporary is returned

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]
This should probably target stable though

**Comment by WalterBright** [comment]
It's marked as critical, but has been there for probably 15 years. It's an ugly bug, but pretty hard to trigger.

File: src/dmd/s2ir.d [hunk 1]
```diff
@@ -687,7 +687,22 @@ private extern (C++) class S2irVisitor : Visitor
...
+                    {
+                         if (VarExp ve = be.e1.isVarExp())
+                         {
+                            if (ve.var == func.vresult)
+                                goto Lskip;
+                         }
+                    }
+
                 e = addressElem(e, s.exp.type.pointerTo());
+             Lskip:
```
Comment (UplinkCoder) [inline src/dmd/s2ir.d:705]: Where is the goto which targets this?

**Comment by UplinkCoder** [comment]
@WalterBright why the self merge?

**Comment by UplinkCoder** [comment]
How does this fix work?

File: src/dmd/s2ir.d [hunk 2]
```diff
@@ -687,7 +687,22 @@ private extern (C++) class S2irVisitor : Visitor
...
+                 */
+                if (func.vresult)
+                    if (BlitExp be = s.exp.isBlitExp())
+                    {
+                         if (VarExp ve = be.e1.isVarExp())
+                         {
+                            if (ve.var == func.vresult)
+                                goto Lskip;
+                         }
+                    }
```
Comment (John-Colvin) [inline src/dmd/s2ir.d:702]: this is the best advert for a `?.` operator or similar I've seen in a while.

```
if (s.exp.isBklitExp()?.e1.isVarExp()?.var == func.vresult)
    goto Lskip;
```

Comment (UplinkCoder) [reply src/dmd/s2ir.d:702]: regardless of `?.`
Even when stated in one line I am still not sure what this actually matches.

**Pull request merged**